README.md
# Alertmanager Forwarder
Alertmanager has built-in integrations for a number of services, but if you need to do more then you're looking at using the Webhook functionality.

This forwarder provides a mechanism for forwarding the alert payload from Alertmanager onto either Graylog or vRealize Orchestrator.  Other integrations could easily be added later as required.

## Graylog configuration
You'll need a suitable UDP input to accept the incoming messages.

Each alert is sent as a separate message with the fields being split out for an easier time at the Graylog end.  The following fields are included in the message payload to Graylog for every event.

 - status
 - startsAt
 - endsAt
 - generatorURL
 - labels
 - annotations

## vRO configuration
Each alert is sent on to vRO as a separate POST request.  The workflow called is based on the workflow ID embedded into the ``vro_action`` field of the alert from Alertmanager.  The alert payload is base64 encoded in order to ensure there are no odd issues with a JSON string being embedded into a JSON string; you'll need to decode the base64 string in your workflow.  Your workflow just needs one input called ``json`` of type ``string``.

## Usage
You don't have to use both Graylog and vRO config options.  If you only include one of them, the forwarder will only function for that one.

| Flag            | Description                                                                         | Default Value |
| --------------- | ------------------------------------------------------------------------------------| ------------- |
| graylog-url     | Graylog listener URL; minus the port                                                | none          |
| graylog-port    | Port that the Graylog UDP input is listening on                                     | 0             |
| vro-host        | vRO listener URL; minus the port                                                    | none          |
| vro-port        | Port that the vRO API is listening on                                               | 0             |
| vro-auth        | base64 encoded authorisation header string.  For example 'dmNvdXNlcjpteXBhc3N3b3Jk' | none          |
| port            | Port that the Alertmanager-Forwarder service listens on                             | 10001         |
| version         | Displays application version information and then quits                             | false         |
| help            | Dsplay usage help                                                                   | false         |
| debug           | Enable debugging                                                                    | false         |
| service         | Manage Windows servers.  Possible options are install, uninstall, start, stop       | none          |

Example command:

````
Alertmanager-Forwarder.exe --service install --graylog-url graylog.mydomain.com --graylog-port 12202 --vro-host vro.mydomain.com --vro-port 8281 --vro-auth dmNvYXBpOmFjcVNaa0V6tGE3bGxQUlRwdDVD
````

This will install the service (not started) using the arguments supplied; it will run on the default port of `10001` as the ``port`` flag was not used..  It will be capable of forwarding alerts to Graylog and vRO.  Omit the ``--service`` flag to run it interactively.

## Downloading a release
<https://github.com/rokett/Alertmanager-Forwarder/releases>

## Building the executable
All dependencies are version controlled, so building the project is really easy.

1. ``go get github.com/rokett/alertmanager-forwarder``.
2. From within the repository directory run ``make build``.
3. Hey presto, you have an executable.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [1.1.0] - 2018-07-24
### Changed
 - #3 Amended to track errors when sending individual alerts to vRO, and return them all once all of the alerts have been iterated over.  This stops one bad alert from causing none of the subsequent alerts being processed.

## [1.0.1] - 2018-07-17
### Fixed
 - #1 --vro-port flag was not being passed to the service configuration when installing.
 - #2 Fixed bad logic whereby not all alerts were being passed to vRO.

## [1.0.0] - 2018-07-16
Initial release


vro.go
package main

import (
	"bytes"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"net/http"

	"github.com/kardianos/service"
)

type vRO struct {
	Parameters []params `json:"parameters"`
}

type params struct {
	Type  string `json:"type"`
	Name  string `json:"name"`
	Scope string `json:"scope"`
	Value val    `json:"value"`
}

type val struct {
	String str `json:"string"`
}

type str struct {
	Value string `json:"value"`
}

type vroError struct {
	Alert map[string]interface{}
	Msg   string
}

func processVRO(logger service.Logger, host string, port int, auth string) http.HandlerFunc {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		switch r.Method {
		case "POST":
			defer r.Body.Close()

			var response response

			var alertGroup alertGroup

			err := json.NewDecoder(r.Body).Decode(&alertGroup)
			if err != nil {
				msg := fmt.Sprintf("Error decoding json body from Alertmanager.\n\n%s", err)
				logger.Error(msg)

				response.Msg = "error decoding json body from AlertManager"
				response.Error = err
				response.Send(http.StatusBadRequest, w)

				return
			}

			var vroErrors []vroError

			for _, alert := range alertGroup.Alerts {
				var workflowID string

				al := mapAlert(alert)

				for k, v := range alert.Labels.(map[string]interface{}) {
					if k == "vro_action" {
						workflowID = v.(string)
						break
					}
				}

				if workflowID == "" {
					msg := fmt.Sprintf("Alert is missing vro_action; skipping.\n\n%s", alert.Labels)
					logger.Error(msg)

					vroErrors = append(vroErrors, vroError{
						Alert: al,
						Msg:   "Alert is missing vro_action; skipping",
					})

					continue
				}

				alertPL, err := json.Marshal(alert)
				if err != nil {
					msg := fmt.Sprintf("Error marshalling alert struct to payload.\n\n%s", err)
					logger.Error(msg)

					vroErrors = append(vroErrors, vroError{
						Alert: al,
						Msg:   msg,
					})

					continue
				}

				encoded := base64.StdEncoding.EncodeToString(alertPL)

				// Yes the JSON structure for vRO API requests is pretty horrid!
				// str -> val -> params -> full payload.  Example below with a truncated base64 encoded value field.
				/*
					{
						"parameters": [
							{
								"type": "string",
								"name": "json",
								"scope": "local",
								"value": {
									"string": {
										"value": "eyJzdGF0dXM"
									}
								}
							}
						]
					}
				*/
				a := str{
					Value: encoded,
				}

				vroVal := val{
					String: a,
				}

				var vroParams []params
				vroParams = append(vroParams, params{
					Type:  "string",
					Name:  "json",
					Scope: "local",
					Value: vroVal,
				})

				vroPL := vRO{
					Parameters: vroParams,
				}

				payload, err := json.Marshal(vroPL)
				if err != nil {
					msg := fmt.Sprintf("Error marshalling vRO struct to payload.\n\n%s", err)
					logger.Error(msg)

					vroErrors = append(vroErrors, vroError{
						Alert: al,
						Msg:   msg,
					})

					continue
				}

				url := fmt.Sprintf("https://%s:%d/vco/api/workflows/%s/executions/", host, port, workflowID)

				req, err := http.NewRequest("POST", url, bytes.NewReader(payload))
				if err != nil {
					msg := fmt.Sprintf("Error creating vRO HTTP request object.\n\nURL: %s\n\n%s", url, err)
					logger.Error(msg)

					vroErrors = append(vroErrors, vroError{
						Alert: al,
						Msg:   msg,
					})

					continue
				}

				authHdr := fmt.Sprintf("Basic %s", auth)
				req.Header.Set("Authorization", authHdr)

				req.Header.Set("Content-Type", "application/json")

				resp, err := http.DefaultClient.Do(req)
				if err != nil {
					msg := fmt.Sprintf("Error POSTing to vRO.\n\n%s", err)
					logger.Error(msg)

					vroErrors = append(vroErrors, vroError{
						Alert: al,
						Msg:   msg,
					})

					continue
				}
				defer resp.Body.Close()

				body, err := ioutil.ReadAll(resp.Body)

				switch resp.StatusCode {
				case 201, 202:
					logger.Info("POST accepted by vRO.\n\nWorkflow execution started.")
				case 400:
					msg := fmt.Sprintf("Unknown error from vRO.\n\nResponse\n%s", string(body))
					logger.Error(msg)

					vroErrors = append(vroErrors, vroError{
						Alert: al,
						Msg:   msg,
					})

					continue
				case 401:
					msg := fmt.Sprint("User not authorised to access vRO API")
					logger.Error(msg)

					vroErrors = append(vroErrors, vroError{
						Alert: al,
						Msg:   msg,
					})

					continue
				default:
					msg := fmt.Sprintf("Unknown status code from vRO.\n\nStatus Code: %d", resp.StatusCode)
					logger.Info(msg)

					vroErrors = append(vroErrors, vroError{
						Alert: al,
						Msg:   msg,
					})

					continue
				}
			}

			if len(vroErrors) > 0 {
				response.Errors = vroErrors
				response.Send(http.StatusInternalServerError, w)
			}
		default:
			msg := fmt.Sprintf("Invalid HTTP method called; %s", r.Method)
			logger.Error(msg)

			http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
			return
		}
	})
}


alert.go
package main

import (
	"time"
)

type alertGroup struct {
	Version           string      `json:"version"`
	GroupKey          string      `json:"groupKey"`
	Status            string      `json:"status"`
	Receiver          string      `json:"receiver"`
	GroupLabels       interface{} `json:"groupLabels"`
	CommonLabels      interface{} `json:"commonLabels"`
	CommonAnnotations interface{} `json:"commonAnnotations"`
	ExternalURL       string      `json:"externalURL"`
	Alerts            []alert     `json:"alerts"`
}

type alert struct {
	Status       string      `json:"status"`
	Labels       interface{} `json:"labels"`
	Annotations  interface{} `json:"annotations"`
	StartsAt     time.Time   `json:"startsAt"`
	EndsAt       time.Time   `json:"endsAt"`
	GeneratorURL string      `json:"generatorURL"`
}


graylog.go
package main

import (
	"encoding/json"
	"fmt"
	"net/http"

	"github.com/kardianos/service"

	"github.com/aphistic/golf"
)

func processGL(logger service.Logger, l *golf.Logger) http.HandlerFunc {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		switch r.Method {
		case "POST":
			var response response

			var alertGroup alertGroup

			err := json.NewDecoder(r.Body).Decode(&alertGroup)
			if err != nil {
				msg := fmt.Sprintf("Error decoding json body from Alertmanager.\n\n%s", err)
				logger.Error(msg)

				response.Msg = "error decoding json body from AlertManager"
				response.Error = err
				response.Send(http.StatusBadRequest, w)

				return
			}

			for _, alert := range alertGroup.Alerts {
				payload := mapAlert(alert)

				l.Infom(payload, "Alert from Alertmanager")
			}
		default:
			msg := fmt.Sprintf("Invalid HTTP method called; %s", r.Method)
			logger.Error(msg)

			http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
			return
		}
	})
}


main.go
package main

import (
	"flag"
	"fmt"
	"log"
	"net/http"
	"os"
	"strconv"

	"github.com/aphistic/golf"

	"github.com/kardianos/service"
)

var (
	app               = "Alertmanager-Forwarder"
	version           string
	build             string
	gl                = flag.String("graylog-url", "", "Define the Graylog listener URL; minus the port")
	glPort            = flag.Int("graylog-port", 0, "Graylog listener port")
	vroHost           = flag.String("vro-host", "", "Define the vRO host used to build the API URL (including the port).  For example 'vco.mydomain.com:8281'")
	vroAuth           = flag.String("vro-auth", "", "Define the base64 encoded authorisation header string.  For example 'dmNvdXNlcjpteXBhc3N3b3Jk'.")
	vroPort           = flag.Int("vro-port", 0, "vRO listener port")
	svrPort           = flag.Int("port", 10001, "Web server port")
	help              = flag.Bool("help", false, "Display help")
	versionFlg        = flag.Bool("version", false, "Display application version")
	debug             = flag.Bool("debug", false, "Enable debugging?")
	winServiceCommand = flag.String("service", "", "Manage Windows services: install, uninstall, start, stop")
	serviceDesc       = "Forward alerts from Alertmanager to another service"
	logger            service.Logger
)

type program struct{}

func main() {
	flag.Parse()

	if *help {
		flag.PrintDefaults()
		os.Exit(0)
	}

	if *versionFlg {
		fmt.Printf("%s v%s build %s", app, version, build)
		os.Exit(0)
	}

	if (*winServiceCommand == "install") && (*gl == "" && *vroHost == "") {
		fmt.Println("A forwarding destination MUST be set")
		flag.PrintDefaults()
		os.Exit(1)
	}

	if *gl != "" && *glPort == 0 {
		fmt.Println("You MUST set the --graylog-port flag")
		flag.PrintDefaults()
		os.Exit(1)
	}

	if *vroHost != "" && *vroPort == 0 {
		fmt.Println("You MUST set the --vro-port flag")
		flag.PrintDefaults()
		os.Exit(1)
	}

	if *vroHost != "" && *vroAuth == "" {
		fmt.Println("You MUST set the --vro-auth flag")
		flag.PrintDefaults()
		os.Exit(1)
	}

	svcConfig := &service.Config{
		Name:        app,
		DisplayName: app,
		Description: serviceDesc,
	}

	prg := &program{}

	svc, err := service.New(prg, svcConfig)
	if err != nil {
		log.Println("Error creating new service")
		log.Println(err)

		os.Exit(1)
	}

	errs := make(chan error, 5)
	logger, err = svc.Logger(errs)
	if err != nil {
		log.Fatal(err)
	}

	go func() {
		for {
			err := <-errs
			if err != nil {
				log.Print(err)
			}
		}
	}()

	if *winServiceCommand != "" {
		if *gl != "" {
			(*svcConfig).Arguments = append((*svcConfig).Arguments, "--graylog-url", *gl)
			(*svcConfig).Arguments = append((*svcConfig).Arguments, "--graylog-port", strconv.Itoa(*glPort))
		}

		if *vroHost != "" {
			(*svcConfig).Arguments = append((*svcConfig).Arguments, "--vro-host", *vroHost)
			(*svcConfig).Arguments = append((*svcConfig).Arguments, "--vro-port", strconv.Itoa(*vroPort))
			(*svcConfig).Arguments = append((*svcConfig).Arguments, "--vro-auth", *vroAuth)
		}

		err := service.Control(svc, *winServiceCommand)
		if err != nil {
			msg := fmt.Sprintf("Error creating new service.\n\n%s", err)
			logger.Error(msg)
		}

		os.Exit(0)
	} else {
		err = svc.Run()
		if err != nil {
			msg := fmt.Sprintf("Error running application interactively.\n\n%s", err)
			logger.Error(msg)
		}
	}
}

func (p *program) Start(s service.Service) error {
	// Start should not block; do the actual work async.
	go p.run(s)
	return nil
}

func (p *program) run(s service.Service) {
	if *gl != "" {
		c, err := golf.NewClient()
		if err != nil {
			msg := fmt.Sprintf("Unable to create new golf client.\n\n%s", err)
			logger.Error(msg)

			p.Stop(s)
		}
		defer c.Close()

		url := fmt.Sprintf("udp://%s:%d", *gl, *glPort)
		err = c.Dial(url)
		if err != nil {
			msg := fmt.Sprintf("Unable to dial Graylog input.\n\n%s", err)
			logger.Error(msg)

			p.Stop(s)
		}

		l, err := c.NewLogger()
		if err != nil {
			msg := fmt.Sprintf("Unable to create new golf logger.\n\n%s", err)
			logger.Error(msg)

			p.Stop(s)
		}

		l.SetAttr("app", app)
		l.SetAttr("app_version", version)
		l.SetAttr("app_build", build)

		http.Handle("/graylog", processGL(logger, l))
	}

	if *vroHost != "" {
		http.Handle("/vro", processVRO(logger, *vroHost, *vroPort, *vroAuth))
	}

	listeningPort := fmt.Sprintf(":%d", *svrPort)

	msg := fmt.Sprintf("Starting %s.  Version %s.  Build %s.\n\nListening on port %s", app, version, build, listeningPort)
	logger.Info(msg)

	err := http.ListenAndServe(listeningPort, nil)
	if err != nil {
		logger.Error(err)

		p.Stop(s)
	}
}

func (p *program) Stop(s service.Service) error {
	logger.Info("Stopping")

	return nil
}
